lib/operations/mongo_client_ops.js

```diff
@@ -600,4 +628,80 @@ function validOptions(options) {
   }
 }
 
+const apmLoggingEvents = {
+  command: ['commandStarted', 'commandSucceeded', 'commandFailed'],
+  sdam: [
+    'serverOpening',
+    'serverDescriptionChanged',
+    'serverHeartbeatStarted',
+    'serverHeartbeatSucceeded',
+    'serverHeartbeatFailed',
+    'serverClosed',
+    'topologyOpening',
+    'topologyClosed',
+    'topologyDescriptionChanged'
+  ]
+};
+
+function apmLoggingEventHandler(client) {
+  return function(err) {
+    if (!err) return;
+    if (!client.topology) throw err;
+    client.close(true, () => {
+      throw err;
+    });
+  };
+}
+
+function getMonitoringCategories(monitor) {
+  const allMonitors = ['command', 'sdam'];
+  // enable all monitors if monitor = all
+  let monitors = monitor === 'all' ? allMonitors : monitor;
+  // ensure `monitors` is an array
+  monitors = Array.isArray(monitors) ? monitors : [monitors];
+  return monitors;
+}
+
+function resolveMonitoringOutputFile(monitorOut) {
+  const defaultFileName = `${process.cwd()}/monitor.log`;
+  if (!monitorOut) return defaultFileName;
+  if (monitorOut === 'stderr' || monitorOut === 'stdout') return monitorOut;
+  return path.resolve(monitorOut);
+}
+
+function createMonitoringStream(fileName) {
+  if (fileName === 'stderr' || fileName === 'stdout') return process[fileName];
+  if (fs.existsSync(fileName)) {
+    const timestamp = new Date().toISOString();
+    fs.renameSync(fileName, `${fileName}-${timestamp}`);
+  }
+  return fs.createWriteStream(fileName);
+}
+
+/**
+ * Attach a category (`command` or `sdam`) of event listeners to the given
+ * client. When an event of that category is emitted, a log line will
+ * be written to the given writable stream.
+ *
+ * @param {MongoClient} client the client to which the listeners are attached
+ * @param {String} category the category of events to listen for
+ * @param {WritableStream} stream stream to which log output will be written
+ */
+function attachMonitorListenerCategory(client, category, stream) {
+  apmLoggingEvents[category].forEach(eventName => {
+    client.on(eventName, e => {
+      const timestamp = new Date().toISOString();
+      const logObj = {
+        timestamp: timestamp,
```

Comment:
it might be easier to use `timestamp: Date.now()` here, since it will return msSinceEpoch which can be easily fed to a `Date` constructor in the future